### PR TITLE
Implement the OAuth2WebViewClient related classes

### PR DESCRIPTION
**Pull request by heidijinxujia**
* Implement the OAuth2WebViewClient #155 
* Implement the NtlmChallengeHandler #156 
* Add the ChallengeCompletionCallback interface #157

### CODE REVIEW

File: common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/OAuth2WebViewClient.java [hunk 1]
```diff
@@ -57,18 +63,121 @@ public Context getContext() {
...
+        // Send the result back to the calling activity
+        mCompletionCallback.sendResponse(AuthenticationConstants.UIResponse.BROWSER_CODE_ERROR, resultIntent);
+    }
+
+    @Override
+    public void onPageFinished(final WebView view,
+                               final String url) {
+        super.onPageFinished(view, url);
+        // Once web view is fully loaded,set to visible
+        view.setVisibility(View.VISIBLE);
```
Comment (iambmelt) [inline common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/OAuth2WebViewClient.java]: Is this view not visible by default?

File: common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/challengehandlers/IChallengeCompletionCallback.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+ * TODO AuthenticationActivity should implement the sendResponse method and call Activity.setResult() and Activity.finish() to return the UI response to the caller.
+ */
+public interface IChallengeCompletionCallback {
+    /**
+     * Send the authorization challenge response back to the activity.
+     *
+     * @param returnCode     UI response code
+     * @param responseIntent challenge response
+     */
+    void sendResponse(int returnCode, Intent responseIntent);
```
Comment (iambmelt) [inline common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/challengehandlers/IChallengeCompletionCallback.java]: Naming-wise, this is _receiving_ the response back, correct?

Would a name like `onChallengeResponseReceived()` be clearer?

File: common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/challengehandlers/IChallengeHandler.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package com.microsoft.identity.common.internal.ui.embeddedwebview.challengehandlers;
+
+public interface IChallengeHandler {
```
Comment (iambmelt) [inline common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/challengehandlers/IChallengeHandler.java]: I'm not sure about the value of this interface -- accepts no params and returns nothing.

Would this make sense to provide with a generic type?

As a thought...

```java
public interface IChallengeHandler<Challenge, Response> {

    Response processChallenge(Challenge challenge);

}
```

and then have....

```java
public final class NtlmChallengeHandler implements IChallengeHandler<NtlmChallenge, Void> {

    @Override
    public Void processChallenge(final NtlmChallenge challenge) {
        if (challenge.getHandler().useHttpAuthUsernamePassword() && challenge.getView() != null) {
        
        final String[] haup = challenge.getView().getHttpAuthUsernamePassword(challenge.getHost(), challenge.mRealm);
            
            if (haup != null && haup.length == 2) {
                final String userName = haup[0];
                final String password = haup[1];
                if (userName != null && password != null) {
                    challenge.getHandler().proceed(userName, password);
                }
            }
        } else {
            showHttpAuthDialog();
        }

        return null;
    }

}
```

File: common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/challengehandlers/IChallengeHandler.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    void process();
```
Comment (iambmelt) [inline common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/challengehandlers/IChallengeHandler.java]: Ideally the method gets javadoc'd at the interface level

File: common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/challengehandlers/NtlmChallengeHandler.java [hunk 5]
```diff
@@ -22,6 +22,137 @@
...
+                            }
+                        }).create().show();
+    }
+
+    private void cancelRequest() {
+        Logger.verbose(TAG, "Sending intent to cancel authentication activity");
+        mChallengeCallback.sendResponse(AuthenticationConstants.UIResponse.BROWSER_CODE_CANCEL, new Intent());
+    }
+
+    class NtlmChallenge {
```
Comment (iambmelt) [inline common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/challengehandlers/NtlmChallengeHandler.java]: Let's make this its own top-level class

File: common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/OAuth2WebViewClient.java [hunk 6]
```diff
@@ -57,18 +63,121 @@ public Context getContext() {
...
+
+    @Override
+    public void onReceivedHttpAuthRequest(WebView view, final HttpAuthHandler handler,
+                                          String host, String realm) {
+        // Create a dialog to ask for credentials and post it to the handler.
+        Logger.info(TAG, "Receive the http auth request. Start the dialog to ask for creds. ");
+        Logger.infoPII(TAG, "Host:" + host);
+
+        //TODO TelemetryEvent.setNTLM(true); after the Telemetry is finished in common.
+        final NtlmChallengeHandler ntlmChallengeHandler
```
Comment (iambmelt) [inline common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/OAuth2WebViewClient.java]: The order of events and object responsibilities feel a little mixed here.

Currently:
The `NtlmChallengeHandler` is created.
The `NtlmChallengeHandler` creates the `NtlmChallenge` (So it's both a handler _and_ a factory)
The `NtlmChallengeHandler` then processes the `NtlmChallenge` that it created. (Then it does some processing)

I suggest we split these actions and responsibilities such that creating the challenge, creating the handler, and doing the processing are 3 discreet, testable, operations - to spitball some pseudocode...

```java
// We have a Factory capable of producing different kinds of challenges
NtlmChallenge ntlmChallenge = new ChallengeFactory().newNtlmChallenge(view, handler, host, realm);

// We init our ChallengeHandler, could conceivably respond to different types of challenges
IChallengeHandler<NtlmChallenge, Void> challengeHandler = new NtlmChallengeHandler();

// We use the ChallengeHandler
challengeHandler.processChallenge(ntlmChallenge);
```

By separating the creation of the `NtlmChallenge` from the `ChallengeHandler` we create a new testable surface in the factory. We can also make the `ChallengeHandler` API more generic by getting rid of the coupling between the Handler and the type of Challenge that's constructed.

File: common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/challengehandlers/NtlmChallengeHandler.java [hunk 7]
```diff
@@ -22,6 +22,137 @@
...
+     * @param host
+     * @param realm
+     * @param context
+     * @param callback
+     */
+    public NtlmChallengeHandler(final WebView view,
+                                final HttpAuthHandler handler,
+                                final String host,
+                                final String realm,
+                                final Context context,
```
Comment (iambmelt) [inline common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/challengehandlers/NtlmChallengeHandler.java]: Is this supposed to be an `Activity` ? Or will _any_ type of `Context` do?

**Comment by heidijinxujia** [comment]
TODO add unit tests

File: common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/OAuth2WebViewClient.java (hunk 1 above)
Comment (heidijinxujia) [reply common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/OAuth2WebViewClient.java]: No, we set the webview `INVISIBLE` when creating it, and set it `VISIBLE` once the webview finish loading.

File: common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/challengehandlers/IChallengeHandler.java (hunk 3 above)
Comment (heidijinxujia) [reply common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/challengehandlers/IChallengeHandler.java]: `Response processChallenge(Challenge challenge);`  could not fit into the `ClientCertAuthChallenge` which will take `ClientCertRequest` (pre-defined class in Android SDK) as input. 

And the challenge result is sent to IChallengeCompletionCallback which required the calling Activity.setResult() and Acitivity.finish() in AuthenticationAcitivty of MSAL. That's why I introduced the callback interface here and let the MSAL to have the callback implementation. This is why there is no return object here.

Comment (heidijinxujia) [reply common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/challengehandlers/IChallengeHandler.java]: Talked to Brian offline. Learned a lot from Brian. I will address this. Thanks @iambmelt

**Comment by heidijinxujia** [comment]
@iambmelt  Thanks a lot for the reviews. I have addressed all your comments. Please let me know if there is any other concerns.

File: common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/OAuth2WebViewClient.java [hunk 8]
```diff
@@ -53,22 +63,130 @@ public AuthorizationRequest getRequest() {
...
+            return;
         }
 
-        mContext = context;
-        mRequest = request;
+        final Uri uri = Uri.parse(url);
+        if (uri.isOpaque()) {
+            Logger.verbose(TAG, "onPageStarted: Non-hierarchical loading uri.");
+            Logger.verbosePII(TAG, "start url: " + url);
+            return;
```
Comment (piqiums) [inline common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/OAuth2WebViewClient.java]: after reading the context, I think making L178-end of the method in a 'if- else if - else' statement would be more readable. A 'return' statement here hurts the readability .

File: common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/challengehandlers/NtlmChallengeHandler.java [hunk 9]
```diff
@@ -22,6 +22,95 @@
...
+    @Override
+    public Void processChallenge(final NtlmChallenge ntlmChallenge) {
+        if (ntlmChallenge.getHandler().useHttpAuthUsernamePassword()
+                && ntlmChallenge.getView() != null) {
+            final String[] haup = ntlmChallenge.getView()
+                    .getHttpAuthUsernamePassword(ntlmChallenge.getHost(), ntlmChallenge.getRealm());
+            if (haup != null && haup.length == 2) {
+                final String userName = haup[0];
+                final String password = haup[1];
+                if (userName != null && password != null) {
```
Comment (piqiums) [inline common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/challengehandlers/NtlmChallengeHandler.java]: for the String 'username' and 'password', checking 'null' is not enough, also please check the length of the String

File: common/src/main/res/layout/activity_authentication.xml [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+    xmlns:tools="http://schemas.android.com/tools"
+   android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    
+    tools:context=".AuthenticationActivity" >
+
+    <WebView
+        android:id="@+id/webView1"
+        android:layout_width="match_parent"
+        android:layout_height="fill_parent"
```
Comment (piqiums) [inline common/src/main/res/layout/activity_authentication.xml]: dah-dah. line 10 is using "match_parent", line 11 is using 'fill_parent'.

File: common/src/main/res/layout/dialog_authentication.xml [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+    android:layout_width="match_parent"
+    android:layout_height="match_parent" >
+
+ 
+
+    <WebView
+        android:id="@+id/com_microsoft_aad_adal_webView1"
+        android:layout_width="fill_parent"
+        android:layout_height="fill_parent"
+        android:paddingBottom="0dp"
```
Comment (piqiums) [inline common/src/main/res/layout/dialog_authentication.xml]: in this case , android:padding="0dp" would be OK

File: common/src/main/res/layout/dialog_authentication.xml [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+        android:paddingTop="0dp" />
+
+    <EditText
+        android:id="@+id/com_microsoft_aad_adal_editDummyText"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_alignBottom="@+id/com_microsoft_aad_adal_webView1"
+        android:layout_alignLeft="@+id/com_microsoft_aad_adal_webView1"
+        android:layout_alignStart="@+id/com_microsoft_aad_adal_webView1"
+        android:layout_marginBottom="124dp"
```
Comment (piqiums) [inline common/src/main/res/layout/dialog_authentication.xml]: this type of dimension constant, please add them into 'dimen.xml'

File: common/src/main/res/layout/dialog_authentication.xml [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+        android:layout_marginLeft="86dp"
+        android:layout_marginStart="86dp"
```
Comment (piqiums) [inline common/src/main/res/layout/dialog_authentication.xml]: ditto

File: common/src/main/res/layout/http_auth_dialog.xml [hunk 14]
```diff
@@ -0,0 +1,31 @@
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:id="@+id/LinearLayout1"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    android:orientation="vertical"
+    android:paddingLeft="10dp"
```
Comment (piqiums) [inline common/src/main/res/layout/http_auth_dialog.xml]: please move to 'dimen.xml'

File: common/src/main/res/layout/http_auth_dialog.xml [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+
+    <EditText
+        android:id="@+id/editPassword"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:ems="10"
+        android:hint="@string/http_auth_dialog_password"
+        android:inputType="textPassword"
+        android:paddingTop="10dp"
+        android:paddingBottom="20dp" />
```
Comment (piqiums) [inline common/src/main/res/layout/http_auth_dialog.xml]: ditto

File: common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/OAuth2WebViewClient.java [hunk 16]
```diff
@@ -53,22 +63,130 @@ public AuthorizationRequest getRequest() {
     /**
      * @return context
      */
-    public Context getContext() {
-        return mContext;
+    public Context getActivity() {
```
Comment (iambmelt) [inline common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/OAuth2WebViewClient.java]: `public Activity getActivity()`

**Comment by heidijinxujia** [comment]
@piqiums @iambmelt  Thank you for the reviews. All comments addressed.

**Pull request merged**